File: moderate/city_blocks.lua
for line in io.lines(arg[1]) do
  local p, q = line:match("%(([%d,]+)%) %(([%d,]+)%)")
  local st, av = {}, {}
  for i in p:gmatch("%d+") do st[#st + 1] = tonumber(i) end
  for i in q:gmatch("%d+") do av[#av + 1] = tonumber(i) end
  if #st < 2 or #av < 2 then
    print(0)
  else
    if st[1] ~= 0 then
      for i = #st, 1, -1 do st[i] = st[i] - st[1] end
    end
    if av[1] ~= 0 then
      for i = #av, 1, -1 do av[i] = av[i] - av[1] end
    end
    local intersec, stl, avl = 0, st[#st], av[#av]
    for i = 1, #av do av[i] = av[i] * stl / avl end
    local i, j = 0, 0
    while i < #st and j < #av do
      if st[i] == av[j] then
        intersec = intersec + 1
        i = i + 1
        j = j + 1
      elseif st[i] < av[j] then
        i = i + 1
      else
        j = j + 1
      end
    end
    print(#st + #av - intersec - 1)
  end
end

File: moderate/city_blocks.hs
import System.Environment (getArgs)
import Data.List.Split (splitOn)
import Data.Ratio (Ratio, (%))

clean   :: String -> String
clean xs = [x | x <- xs, notElem x "()"]

cityBlock                       :: Int -> [Ratio Int] -> [Ratio Int] -> Int
cityBlock intersec []     _      = intersec + 1
cityBlock intersec _      []     = intersec + 1
cityBlock intersec (x:xs) (y:ys) | x == y    = cityBlock (succ intersec) xs ys
                                 | x < y     = cityBlock intersec xs (y:ys)
                                 | otherwise = cityBlock intersec (x:xs) ys

cityBlocks         :: [String] -> Int
cityBlocks [xs, ys] | length ps < 2 || length qs < 2 = 0
                    | otherwise                      = length ps + length qs - cityBlock 0 ps'' qs''
                    where ps   = map (read :: String -> Int) (splitOn "," xs)
                          qs   = map (read :: String -> Int) (splitOn "," ys)
                          ps'  = map (+ (-(head ps))) ps
                          qs'  = map (+ (-(head qs))) qs
                          stl  = last ps'
                          avl  = last qs'
                          ps'' = map fromIntegral ps'
                          qs'' = map (\x -> (x * stl) % avl) qs'

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map (show . cityBlocks . words) . lines $ clean input
